### PR TITLE
mm: Enable a dedicated kernel heap on `BUILD_FLAT` via `MM_KERNEL_HEAP`

### PR DESCRIPTION
**Pull request by gustavonihei**
## Summary

This PR intends to enable a dedicated kernel heap on NuttX **Flat mode** by extending `MM_KERNEL_HEAP` feature.

By separating the kernel and userspace heaps, the user is granted more control over the heaps placement within the memory hierarchy, which is specially useful for microcontrollers that provide External RAM.

Besides segregating the kernel and userspace allocations, this feature does not prevent the userspace from accessing the kernel heap under `BUILD_FLAT`.

`MM_KERNEL_HEAP` support under `BUILD_PROTECTED` and `BUILD_KERNEL` should be unaffected.

## Impact

NuttX memory management subsystem is now extended with a new feature, enabled only on specific combination of Kconfig options.

## Testing

Successful execution of `ostest` in `esp32-devkitc:psram_usrheap` defconfig, to be submitted later.

### CODE REVIEW

File: include/nuttx/mm/mm.h
Comment (xiaoxiang781216) [inline include/nuttx/mm/mm.h:92]:
```suggestion
#if !defined(__KERNEL__) && (defined(CONFIG_BUILD_FLAT)
```

Comment (gustavonihei) [reply include/nuttx/mm/mm.h:92]: This suggestion is not equivalent. Is this what you intended to suggest?

```suggestion
#if !defined(__KERNEL__) && !defined(CONFIG_BUILD_FLAT)
```

Comment (xiaoxiang781216) [reply include/nuttx/mm/mm.h:92]: or merge into one line:
```
#if defined(__KERNEL__) && defined(CONFIG_BUILD_KERNEL) || defined(CONFIG_BUILD_PROTECTED)
```

Comment (xiaoxiang781216) [reply include/nuttx/mm/mm.h:92]: Sure, it's fine but we need indent `#ifndef __KERNEL` two spaces

Comment (acassis) [reply include/nuttx/mm/mm.h:92]: Gustavo, maybe it could be a good idea to put that comment:

CONFIG_MM_KERNEL_HEAP must be undefined only when building the Userspace image, which only exists under CONFIG_BUILD_KERNEL and CONFIG_BUILD_PROTECTED. So we need to ensure !defined(__KERNEL__)

 just before the #if testing to let people understand easily the intention

**Review by acassis — APPROVED** [review]

File: include/nuttx/mm/mm.h
Comment (gustavonihei) [reply include/nuttx/mm/mm.h]: Thanks @xiaoxiang781216 and @acassis for the feedback, I've implemented both suggestions. PTAL.

**Comment by gustavonihei** [comment]
CI failure is unrelated to this PR. Prior to rebasing it passed successfully, should we proceed with the merge or wait until #8155 is merged?

**Comment by xiaoxiang781216** [comment]
let's ignore the unrelated ci broken.

**Pull request merged**
